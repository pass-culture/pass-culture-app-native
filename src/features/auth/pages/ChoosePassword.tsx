import { t } from '@lingui/macro'
import { useNavigation } from '@react-navigation/native'
import { StackScreenProps } from '@react-navigation/stack'
import React, { FunctionComponent, useState } from 'react'
import { Alert } from 'react-native'
import styled from 'styled-components/native'

import {
  isPasswordCorrect,
  PasswordSecurityRules,
} from 'features/auth/components/PasswordSecurityRules'
import { RootStackParamList } from 'features/navigation/RootNavigator'
import { _ } from 'libs/i18n'
import { BottomCard, BottomCardContentContainer } from 'ui/components/BottomCard'
import { ButtonPrimary } from 'ui/components/buttons/ButtonPrimary'
import { PasswordInput } from 'ui/components/inputs/PasswordInput'
import { ModalHeader } from 'ui/components/modals/ModalHeader'
import { Background } from 'ui/svg/Background'
import { ArrowPrevious } from 'ui/svg/icons/ArrowPrevious'
import { Close } from 'ui/svg/icons/Close'
import { getSpacing, Spacer, Typo } from 'ui/theme'

type Props = StackScreenProps<RootStackParamList, 'ChoosePassword'>

export const ChoosePassword: FunctionComponent<Props> = ({ navigation }) => {
  const [password, setPassword] = useState('')
  const { goBack } = useNavigation()

  // TODO: PC-5430 gestion du storage de l'email (props ? navigation params ?) & transmission password

  function onClose() {
    Alert.alert('TODO: PC-4936 abandon registration')
  }

  function submitPassword() {
    navigation.navigate('SetBirthday')
  }

  return (
    <React.Fragment>
      <Background />
      <BottomCard>
        <ModalHeader
          title={_(t`Ton mot de passe`)}
          rightIcon={Close}
          onRightIconPress={onClose}
          leftIcon={ArrowPrevious}
          onLeftIconPress={goBack}
        />
        <BottomCardContentContainer>
          <Spacer.Column numberOfSpaces={6} />
          <StyledInput>
            <Typo.Body>{_(t`Mot de passe`)}</Typo.Body>
            <Spacer.Column numberOfSpaces={2} />
            <PasswordInput
              value={password}
              autoFocus
              onChangeText={setPassword}
              placeholder={_(/*i18n: password placeholder */ t`Ton mot de passe`)}
            />
          </StyledInput>
          <PasswordSecurityRules password={password} />
          <Spacer.Column numberOfSpaces={6} />
          <ButtonPrimary
            title={_(t`Continuer`)}
            onPress={submitPassword}
            disabled={!isPasswordCorrect(password)}
          />
        </BottomCardContentContainer>
      </BottomCard>
    </React.Fragment>
  )
}

const StyledInput = styled.View({
  display: 'flex',
  flexDirection: 'column',
  alignItems: 'flex-start',
  width: '100%',
  maxWidth: getSpacing(125),
})
