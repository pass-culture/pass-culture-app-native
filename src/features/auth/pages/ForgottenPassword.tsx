import { t } from '@lingui/macro'
import { StackScreenProps } from '@react-navigation/stack'
import React, { FunctionComponent, useState } from 'react'
import { Alert } from 'react-native'
import styled from 'styled-components/native'

import { requestPasswordReset } from 'features/auth/api'
import { HomeStackParamList } from 'features/home/navigation/HomeNavigator'
import { _ } from 'libs/i18n'
import { BottomCard } from 'ui/components/BottomCard'
import { ButtonPrimary } from 'ui/components/buttons/ButtonPrimary'
import { isValueEmpty } from 'ui/components/inputs/helpers'
import { TextInput } from 'ui/components/inputs/TextInput'
import { ModalHeader } from 'ui/components/modals/ModalHeader'
import { SafeContainer } from 'ui/components/SafeContainer'
import { ArrowPrevious } from 'ui/svg/icons/ArrowPrevious'
import { Close } from 'ui/svg/icons/Close'
import { getSpacing, Spacer, Typo } from 'ui/theme'

type Props = StackScreenProps<HomeStackParamList, 'ForgottenPassword'>

export const ForgottenPassword: FunctionComponent<Props> = ({ navigation }) => {
  const [email, setEmail] = useState('')

  const shouldDisableValidateButton = isValueEmpty(email)

  function onBackNavigation() {
    navigation.navigate('Login')
  }

  function onClose() {
    navigation.navigate('Home', { shouldDisplayLoginModal: false })
  }

  async function validateEmail() {
    await requestPasswordReset({ email })
      .then(() => {
        navigation.navigate('ResetPasswordEmailSent', { email })
      })
      .catch((err) => {
        Alert.alert(err.message)
      })
  }

  return (
    <SafeContainer noTabBarSpacing>
      <BottomCard>
        <ModalHeader
          title={_(t`Mot de passe oublié`)}
          leftIcon={ArrowPrevious}
          onLeftIconPress={onBackNavigation}
          rightIcon={Close}
          onRightIconPress={onClose}
        />
        <ModalContent>
          <CenteredText>
            <Typo.Body>
              {_(
                t`Saisis ton adresse e-mail pour recevoir un lien qui te permettra de réinitialiser ton mot de passe !`
              )}
            </Typo.Body>
          </CenteredText>
          <Spacer.Column numberOfSpaces={4} />
          <StyledInput>
            <Typo.Body>{_(t`Adresse e-mail`)}</Typo.Body>
            <Spacer.Column numberOfSpaces={2} />
            <TextInput
              value={email}
              onChangeText={setEmail}
              placeholder={_(/*i18n: email placeholder */ t`tonadresse@email.com`)}
              keyboardType="email-address"
              textContentType="emailAddress"
              autoFocus={true}
            />
          </StyledInput>
          <Spacer.Column numberOfSpaces={6} />
          <ButtonPrimary
            title={_(t`Valider`)}
            onPress={validateEmail}
            disabled={shouldDisableValidateButton}
          />
        </ModalContent>
      </BottomCard>
    </SafeContainer>
  )
}

const ModalContent = styled.View({
  paddingTop: getSpacing(7),
  alignItems: 'center',
  width: '100%',
})

const CenteredText = styled.Text({
  textAlign: 'center',
})

const StyledInput = styled.View({
  display: 'flex',
  flexDirection: 'column',
  alignItems: 'flex-start',
  width: '100%',
  maxWidth: getSpacing(125),
})
