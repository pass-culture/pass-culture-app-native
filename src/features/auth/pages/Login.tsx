import { t } from '@lingui/macro'
import { StackScreenProps } from '@react-navigation/stack'
import React, { FunctionComponent, useState } from 'react'
import { Alert } from 'react-native'
import styled from 'styled-components/native'

import { RootStackParamList } from 'features/navigation/RootNavigator'
import { env } from 'libs/environment'
import { _ } from 'libs/i18n'
import { BottomCard } from 'ui/components/BottomCard'
import { ButtonPrimary } from 'ui/components/buttons/ButtonPrimary'
import { isValueEmpty } from 'ui/components/inputs/helpers'
import { PasswordInput } from 'ui/components/inputs/PasswordInput'
import { TextInput } from 'ui/components/inputs/TextInput'
import { ModalHeader } from 'ui/components/modals/ModalHeader'
import { SafeContainer } from 'ui/components/SafeContainer'
import { ArrowPrevious } from 'ui/svg/icons/ArrowPrevious'
import { Close } from 'ui/svg/icons/Close'
import { Warning } from 'ui/svg/icons/Warning'
import { ColorsEnum, getSpacing, Spacer, Typo } from 'ui/theme'

import { signin } from '../api'

type Props = StackScreenProps<RootStackParamList, 'Login'>

let INITIAL_IDENTIFIER = ''
let INITIAL_PASSWORD = ''

if (__DEV__) {
  INITIAL_IDENTIFIER = env.SIGNIN_IDENTIFIER
  INITIAL_PASSWORD = env.SIGNIN_PASSWORD
}

export const Login: FunctionComponent<Props> = function (props: Props) {
  const [email, setEmail] = useState(INITIAL_IDENTIFIER)
  const [password, setPassword] = useState(INITIAL_PASSWORD)
  const [shouldShowErrorMessage, setShouldShowErrorMessage] = useState(false)

  const shouldDisableLoginButton = isValueEmpty(email) || isValueEmpty(password)

  async function handleSignin() {
    setShouldShowErrorMessage(false)
    const isSigninSuccessful = await signin({ email, password })
    if (isSigninSuccessful) {
      props.navigation.navigate('Home', { shouldDisplayLoginModal: false })
    } else {
      setShouldShowErrorMessage(true)
    }
  }

  function onBackNavigation() {
    props.navigation.navigate('Home', { shouldDisplayLoginModal: true })
  }

  function onClose() {
    props.navigation.navigate('Home', { shouldDisplayLoginModal: false })
  }

  function onForgottenPasswordClick() {
    Alert.alert('TO DO')
  }

  return (
    <SafeContainer>
      <BottomCard>
        <ModalHeader
          title={_(t`Connecte-toi !`)}
          leftIcon={ArrowPrevious}
          onLeftIconPress={onBackNavigation}
          rightIcon={Close}
          onRightIconPress={onClose}
        />
        {shouldShowErrorMessage && (
          <React.Fragment>
            <Spacer.Column numberOfSpaces={5} />
            <StyledInline>
              <Warning size={24} />
              <Typo.Caption color={ColorsEnum.ERROR}>
                {_(t`E-mail ou mot de passe incorrect`)}
              </Typo.Caption>
            </StyledInline>
          </React.Fragment>
        )}
        <Spacer.Column numberOfSpaces={7} />
        <StyledInput>
          <Typo.Body>{_(t`Adresse e-mail`)}</Typo.Body>
          <Spacer.Column numberOfSpaces={2} />
          <TextInput
            value={email}
            onChangeText={setEmail}
            placeholder={_(/*i18n: email placeholder */ t`tonadresse@email.com`)}
            keyboardType="email-address"
            autoFocus={true}
            isError={shouldShowErrorMessage}
            width="100%"
          />
        </StyledInput>
        <Spacer.Column numberOfSpaces={6} />
        <StyledInput>
          <Typo.Body>{_(t`Mot de passe`)}</Typo.Body>
          <Spacer.Column numberOfSpaces={2} />
          <PasswordInput
            value={password}
            onChangeText={setPassword}
            placeholder={_(/*i18n: password placeholder */ t`Ton mot de passe`)}
            isError={shouldShowErrorMessage}
            width="100%"
          />
        </StyledInput>
        <Spacer.Column numberOfSpaces={7} />
        <StyledForgottenPasswordTouchableOpacity onPress={onForgottenPasswordClick}>
          <Typo.ButtonText>{_(t`Mot de passe oubli√© ?`)}</Typo.ButtonText>
        </StyledForgottenPasswordTouchableOpacity>
        <Spacer.Column numberOfSpaces={8} />
        <ButtonPrimary
          title={_(t`Se connecter`)}
          onPress={handleSignin}
          disabled={shouldDisableLoginButton}
        />
      </BottomCard>
    </SafeContainer>
  )
}

const StyledInput = styled.View({
  display: 'flex',
  flexDirection: 'column',
  alignItems: 'flex-start',
  width: '100%',
  maxWidth: getSpacing(125),
})

const StyledInline = styled.View({
  display: 'flex',
  flexDirection: 'row',
  justifyContent: 'space-around',
  alignItems: 'center',
  maxWidth: getSpacing(125),
})

const StyledForgottenPasswordTouchableOpacity = styled.TouchableOpacity({
  flexDirection: 'row',
  justifyContent: 'flex-end',
  width: '100%',
  maxWidth: getSpacing(125),
})
