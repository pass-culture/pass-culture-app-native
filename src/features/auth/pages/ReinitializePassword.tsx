import { t } from '@lingui/macro'
import { useNavigation, useRoute } from '@react-navigation/native'
import React, { useEffect, useState } from 'react'
import styled from 'styled-components/native'

import { UseNavigationType, UseRouteType } from 'features/navigation/RootNavigator'
import { _ } from 'libs/i18n'
import { BottomCard } from 'ui/components/BottomCard'
import { ButtonPrimary } from 'ui/components/buttons/ButtonPrimary'
import { PasswordInput } from 'ui/components/inputs/PasswordInput'
import { ModalHeader } from 'ui/components/modals/ModalHeader'
import { SafeContainer } from 'ui/components/SafeContainer'
import { Close } from 'ui/svg/icons/Close'
import { Warning } from 'ui/svg/icons/Warning'
import { ColorsEnum, getSpacing, Spacer, Typo } from 'ui/theme'

const MILLISECONDS_IN_A_SECOND = 1000

export const ReinitializePassword = () => {
  const {
    params: { expiration_date },
  } = useRoute<UseRouteType<'ReinitializePassword'>>()
  const navigation = useNavigation<UseNavigationType>()

  const [password, setPassword] = useState('')
  const [shouldShowPasswordError] = useState(false)
  const [confirmedPassword, setConfirmedPassword] = useState('')
  const [shouldShowConfirmationError] = useState(false)

  const arePasswordsMatching = confirmedPassword === password
  const allowSubmission = password.length > 0 && arePasswordsMatching

  function onClose() {
    navigation.navigate('Home', { shouldDisplayLoginModal: false })
  }
  function submitPassword() {
    // TODO: call submit function using the 'token' params
  }

  useEffect(() => {
    if (expiration_date * MILLISECONDS_IN_A_SECOND < new Date().getTime()) {
      navigation.navigate('Login')
    }
  }, [])

  return (
    <SafeContainer>
      <BottomCard>
        <ModalHeader title={_(t`Ton mot de passe`)} rightIcon={Close} onRightIconPress={onClose} />
        <Spacer.Column numberOfSpaces={6} />
        <StyledInput>
          <Typo.Body>{_(t`Nouveau mot de passe`)}</Typo.Body>
          <Spacer.Column numberOfSpaces={2} />
          <PasswordInput
            value={password}
            onChangeText={setPassword}
            placeholder={_(/*i18n: password placeholder */ t`Ton mot de passe`)}
            isError={shouldShowPasswordError}
            width="100%"
          />
        </StyledInput>
        <Spacer.Column numberOfSpaces={6} />
        <StyledInput>
          <Typo.Body>{_(t`Confirmer le mot de passe`)}</Typo.Body>
          <Spacer.Column numberOfSpaces={2} />
          <PasswordInput
            value={confirmedPassword}
            onChangeText={setConfirmedPassword}
            placeholder={_(/*i18n: password placeholder */ t`Ton mot de passe`)}
            isError={shouldShowConfirmationError}
            width="100%"
          />
        </StyledInput>
        <Spacer.Column numberOfSpaces={2} />
        {!arePasswordsMatching && (
          <ErrorLineContainer>
            <Warning size={24} color={ColorsEnum.ERROR} />
            <Typo.Caption color={ColorsEnum.ERROR}>
              {_(t`les mots de passe ne concordent pas`)}
            </Typo.Caption>
          </ErrorLineContainer>
        )}
        <Spacer.Column numberOfSpaces={6} />
        <ButtonPrimary
          title={_(t`Continuer`)}
          onPress={submitPassword}
          disabled={!allowSubmission}
        />
      </BottomCard>
    </SafeContainer>
  )
}

const StyledInput = styled.View({
  display: 'flex',
  flexDirection: 'column',
  alignItems: 'flex-start',
  width: '100%',
  maxWidth: getSpacing(125),
})

const ErrorLineContainer = styled.View({
  flexDirection: 'row',
  alignItems: 'center',
  justifyContent: 'flex-start',
  width: '100%',
})
