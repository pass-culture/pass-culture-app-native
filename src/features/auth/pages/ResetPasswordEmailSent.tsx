import { t } from '@lingui/macro'
import { StackScreenProps } from '@react-navigation/stack'
import React, { FunctionComponent } from 'react'
import { Alert, Linking } from 'react-native'
import { openInbox } from 'react-native-email-link'
import styled from 'styled-components/native'

import { HomeStackParamList } from 'features/home/navigation/HomeNavigator'
import { _ } from 'libs/i18n'
import { BottomCard } from 'ui/components/BottomCard'
import { ButtonPrimary } from 'ui/components/buttons/ButtonPrimary'
import { ButtonTertiary } from 'ui/components/buttons/ButtonTertiary'
import { ModalHeader } from 'ui/components/modals/ModalHeader'
import { SafeContainer } from 'ui/components/SafeContainer'
import { ArrowPrevious } from 'ui/svg/icons/ArrowPrevious'
import { Close } from 'ui/svg/icons/Close'
import { Email } from 'ui/svg/icons/Email'
import { ExternalSite } from 'ui/svg/icons/ExternalSite'
import { padding, Spacer, Typo } from 'ui/theme'

const SUPPORT_EMAIL_ADDRESS = 'support@passculture.app'

type Props = StackScreenProps<HomeStackParamList, 'ResetPasswordEmailSent'>

export const ResetPasswordEmailSent: FunctionComponent<Props> = ({ navigation, route }) => {
  function onBackNavigation() {
    Alert.alert('TODO => PC-4356')
  }

  async function contactSupport() {
    const url = `mailto:${SUPPORT_EMAIL_ADDRESS}`
    const canOpen = await Linking.canOpenURL(url)

    if (canOpen) {
      Linking.openURL(url)
    }
  }

  function onClose() {
    navigation.navigate('Home', { shouldDisplayLoginModal: false })
  }

  return (
    <SafeContainer noTabBarSpacing>
      <BottomCard>
        <ModalHeader
          title={_(t`E-mail envoyé !`)}
          leftIcon={ArrowPrevious}
          onLeftIconPress={onBackNavigation}
          rightIcon={Close}
          onRightIconPress={onClose}
        />
        <ModalContent>
          <Description>
            <Typo.Body>{_(t`Clique sur le lien reçu à l'adresse :`)}</Typo.Body>
            <Typo.Body>{route.params.email}</Typo.Body>
            <Spacer.Column numberOfSpaces={5} />
            <CenteredText>
              <Typo.Body>
                {_(t`L'e-mail peut prendre jusqu'à 24h pour arriver. Pense à vérifier tes spams !`)}
              </Typo.Body>
            </CenteredText>
            <Spacer.Column numberOfSpaces={5} />
            <Typo.Body>{_(t`S'il n'arrive pas tu peux :`)}</Typo.Body>
            <ButtonTertiary
              title={_(t`Contacter le support`)}
              onPress={contactSupport}
              icon={Email}
            />
            <Spacer.Column numberOfSpaces={6} />
          </Description>
          <ButtonPrimary
            title={_(t`Consulter mes e-mails`)}
            onPress={openInbox}
            icon={ExternalSite}
          />
        </ModalContent>
      </BottomCard>
    </SafeContainer>
  )
}

const ModalContent = styled.View({
  ...padding(4, 1),
  alignItems: 'center',
  width: '100%',
})

const Description = styled.View({
  alignItems: 'center',
})

const CenteredText = styled.Text({
  textAlign: 'center',
})
