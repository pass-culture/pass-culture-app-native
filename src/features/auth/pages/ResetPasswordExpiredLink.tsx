import { t } from '@lingui/macro'
import { StackScreenProps } from '@react-navigation/stack'
import React from 'react'
import { Alert } from 'react-native'
import styled from 'styled-components/native'

import {
  HomeStackParamList,
  navigateToHomeWithoutModal,
} from 'features/home/navigation/HomeNavigator'
import { _ } from 'libs/i18n'
import { ButtonPrimaryWhite } from 'ui/components/buttons/ButtonPrimaryWhite'
import { ButtonTertiaryWhite } from 'ui/components/buttons/ButtonTertiaryWhite'
import { GenericInfoPage } from 'ui/components/GenericInfoPage'
import { Email } from 'ui/svg/icons/Email'
import { SadFace } from 'ui/svg/icons/SadFace'
import { ColorsEnum, Spacer, Typo } from 'ui/theme'

import { requestPasswordReset } from '../api'

import { contactSupport } from './support.services'

type Props = StackScreenProps<HomeStackParamList, 'ResetPasswordExpiredLink'>

export function ResetPasswordExpiredLink(props: Props) {
  async function resendEmailForResetPassword() {
    const { email } = props.route.params
    await requestPasswordReset({ email })
      .then(() => {
        props.navigation.navigate('ResetPasswordEmailSent', { email })
      })
      .catch((error) => {
        Alert.alert(error.message)
      })
  }

  return (
    <GenericInfoPage title={_(t`Oups`)} icon={SadFace}>
      <StyledBody>{_(t`Le lien est expiré !`)}</StyledBody>
      <StyledBody>
        {_(t`Clique sur « Renvoyer l’e-mail » pour recevoir un nouveau lien.`)}
      </StyledBody>
      <Spacer.Column numberOfSpaces={4} />
      <StyledBody>{_(t`Si tu as besoin d’aide n’hésite pas à :`)}</StyledBody>
      <Spacer.Column numberOfSpaces={4} />
      <ButtonTertiaryWhite
        title={_(t`Contacter le support`)}
        onPress={contactSupport}
        icon={Email}
      />
      <Spacer.Column numberOfSpaces={4} />
      <ButtonPrimaryWhite title={_(t`Renvoyer l'email`)} onPress={resendEmailForResetPassword} />
      <Spacer.Column numberOfSpaces={4} />
      <ButtonTertiaryWhite
        title={_(t`Retourner à l'accueil`)}
        onPress={navigateToHomeWithoutModal}
      />
    </GenericInfoPage>
  )
}

const StyledBody = styled(Typo.Body).attrs({
  color: ColorsEnum.WHITE,
})({
  textAlign: 'center',
})
