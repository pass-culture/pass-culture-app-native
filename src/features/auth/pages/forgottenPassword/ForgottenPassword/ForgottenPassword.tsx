import { useNavigation } from '@react-navigation/native'
import React, { useCallback, useMemo } from 'react'
import { useForm } from 'react-hook-form'
import { UseQueryResult } from 'react-query'

import { api } from 'api/api'
import { ApiError } from 'api/ApiError'
import { isAPIExceptionCapturedAsInfo } from 'api/apiHelpers'
import { SettingsResponse } from 'api/gen'
import { useSettingsContext } from 'features/auth/context/SettingsContext'
import { StepperOrigin, UseNavigationType } from 'features/navigation/RootNavigator/types'
import { captureMonitoringError } from 'libs/monitoring'
import { useNetInfoContext } from 'libs/network/NetInfoWrapper'
import { ReCaptchaError } from 'libs/recaptcha/errors'
import { ReCaptcha } from 'libs/recaptcha/ReCaptcha'
import { EmailInputController } from 'shared/forms/controllers/EmailInputController'
import { ButtonPrimary } from 'ui/components/buttons/ButtonPrimary'
import { Form } from 'ui/components/Form'
import { isEmailValid } from 'ui/components/inputs/emailCheck'
import { isValueEmpty } from 'ui/components/inputs/helpers'
import { SecondaryPageWithBlurHeader } from 'ui/pages/SecondaryPageWithBlurHeader'
import { Spacer, Typo } from 'ui/theme'
import { getHeadingAttrs } from 'ui/theme/typographyAttrs/getHeadingAttrs'

type FormValues = {
  email: string
  isFetching: boolean
  isDoingReCaptchaChallenge: boolean
  network: void
  reCaptcha: void
}

export const ForgottenPassword = () => {
  const { data: settings, isLoading: areSettingsLoading } = useSettingsContext()

  const {
    control,
    isDoingReCaptchaChallenge,
    isFetching,
    onBackNavigation,
    onReCaptchaClose,
    onReCaptchaError,
    onReCaptchaExpire,
    onReCaptchaSuccess,
    openReCaptchaChallenge,
    requestPasswordReset,
    shouldDisableValidateButton,
  } = useForgottenPasswordForm(settings)

  return (
    <SecondaryPageWithBlurHeader
      title="Oubli de mot de passe"
      shouldDisplayBackButton
      onGoBack={onBackNavigation}>
      {settings?.isRecaptchaEnabled ? (
        <ReCaptcha
          onClose={onReCaptchaClose}
          onError={onReCaptchaError}
          onExpire={onReCaptchaExpire}
          onSuccess={onReCaptchaSuccess}
          isVisible={isDoingReCaptchaChallenge}
        />
      ) : null}
      <Typo.Title3 {...getHeadingAttrs(2)}>Mot de passe oublié&nbsp;?</Typo.Title3>
      <Spacer.Column numberOfSpaces={2} />
      <Typo.Body>
        Saisis ton adresse e-mail pour recevoir un lien qui te permettra de réinitialiser ton mot de
        passe&nbsp;!
      </Typo.Body>
      <Spacer.Column numberOfSpaces={8} />
      <Form.MaxWidth>
        <EmailInputController control={control} name="email" autoFocus />
        <Spacer.Column numberOfSpaces={8} />
        <ButtonPrimary
          wording="Valider"
          onPress={settings?.isRecaptchaEnabled ? openReCaptchaChallenge : requestPasswordReset}
          isLoading={isDoingReCaptchaChallenge || isFetching || areSettingsLoading}
          disabled={shouldDisableValidateButton}
        />
      </Form.MaxWidth>
    </SecondaryPageWithBlurHeader>
  )
}

const useForgottenPasswordForm = (settings: UseQueryResult<SettingsResponse, unknown>['data']) => {
  const { navigate } = useNavigation<UseNavigationType>()
  const {
    control,
    setError,
    clearErrors,
    setValue,
    watch,
    formState: { errors },
  } = useForm<FormValues>({
    defaultValues: { email: '', isFetching: false, isDoingReCaptchaChallenge: false },
  })
  const { email, isFetching, isDoingReCaptchaChallenge } = watch()

  // Little helper method to make it easier to set error
  const setCustomError = useCallback(
    (message: string) => {
      return setError('email', { type: 'custom', message })
    },
    [setError]
  )

  const onConnection = useCallback(clearErrors, [clearErrors])
  const onConnectionLost = useCallback(() => {
    setCustomError('Hors connexion\u00a0: en attente du réseau.')
    setValue('isDoingReCaptchaChallenge', false)
  }, [setCustomError, setValue])

  const networkInfo = useNetInfoContext({
    onConnection,
    onConnectionLost,
  })
  const { replace } = useNavigation<UseNavigationType>()

  /**
   * Called after recaptcha test, or if recaptcha not needed.
   *
   * @param {string} token Initialized with `dummyToken` by default, since token
   * needs to be a non-empty string even when ReCaptcha validation is deactivated.
   * Cf. backend logic for token validation
   */
  const requestPasswordReset = useCallback(
    async function requestPasswordReset(token = 'dummyToken') {
      clearErrors()
      try {
        setValue('isFetching', true)
        await api.postNativeV1RequestPasswordReset({
          email,
          token: settings?.isRecaptchaEnabled ? token : undefined,
        })
        replace('ResetPasswordEmailSent', { email })
      } catch (error) {
        setCustomError('Un problème est survenu pendant la réinitialisation, réessaie plus tard.')
        if (error instanceof ApiError && !isAPIExceptionCapturedAsInfo(error.statusCode)) {
          captureMonitoringError(error.message, 'ForgottenPasswordRequestResetError')
        }
      } finally {
        setValue('isFetching', false)
      }
    },
    [clearErrors, email, replace, setCustomError, setValue, settings]
  )

  const openReCaptchaChallenge = useCallback(
    function openReCaptchaChallenge() {
      if (!isEmailValid(email)) {
        return setCustomError(
          'L’e-mail renseigné est incorrect. Exemple de format attendu\u00a0: edith.piaf@email.fr'
        )
      }
      if (!networkInfo.isConnected) {
        return setCustomError('Hors connexion\u00a0: en attente du réseau.')
      }
      setValue('isDoingReCaptchaChallenge', true)
      clearErrors()
    },
    [clearErrors, email, networkInfo.isConnected, setCustomError, setValue]
  )

  const errorValues = Object.values(errors)

  return useMemo(
    () => ({
      control,
      openReCaptchaChallenge,
      requestPasswordReset,
      shouldDisableValidateButton: isValueEmpty(email) || isFetching,
      hasError: Boolean(errorValues.length),
      lastError: errorValues[errorValues.length - 1]?.message,
      isFetching,
      isDoingReCaptchaChallenge,
      async onReCaptchaSuccess(token: string) {
        setValue('isDoingReCaptchaChallenge', false)
        await requestPasswordReset(token)
      },
      onReCaptchaExpire() {
        setValue('isDoingReCaptchaChallenge', false)
        setCustomError('Le token reCAPTCHA a expiré, tu peux réessayer.')
      },
      onReCaptchaClose() {
        setValue('isDoingReCaptchaChallenge', false)
      },
      onReCaptchaError(errorCode: ReCaptchaError, error: string | undefined) {
        setValue('isDoingReCaptchaChallenge', false)
        if (errorCode === 'ReCaptchaNetworkError') {
          setCustomError(
            'Un problème est survenu pendant la réinitialisation, vérifie ta connexion internet et réessaie plus tard.'
          )
        } else {
          setCustomError('Un problème est survenu pendant la réinitialisation, réessaie plus tard.')
          captureMonitoringError(`${errorCode} ${error}`, 'ForgottenPasswordOnRecaptchaError')
        }
      },
      onBackNavigation() {
        navigate('Login', { from: StepperOrigin.FORGOTTEN_PASSWORD })
      },
    }),
    [
      control,
      email,
      errorValues,
      isDoingReCaptchaChallenge,
      isFetching,
      navigate,
      openReCaptchaChallenge,
      requestPasswordReset,
      setCustomError,
      setValue,
    ]
  )
}
