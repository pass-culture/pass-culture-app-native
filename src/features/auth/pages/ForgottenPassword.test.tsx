import { StackScreenProps } from '@react-navigation/stack'
import { fireEvent, render } from '@testing-library/react-native'
import React from 'react'
import waitForExpect from 'wait-for-expect'

import { ForgottenPassword } from 'features/auth/pages/ForgottenPassword'
import { HomeStackParamList } from 'features/home/navigation/HomeNavigator'
import { navigationTestProps } from 'tests/navigation'

beforeEach(() => {
  jest.resetAllMocks()
})

jest.mock('@react-navigation/native', () => jest.requireActual('@react-navigation/native'))

describe('<ForgottenPassword />', () => {
  it('should enable validate button when email input is filled', async () => {
    const { getByPlaceholderText, toJSON } = renderPage()
    const disabledButtonSnapshot = toJSON()

    const emailInput = getByPlaceholderText('tonadresse@email.com')
    fireEvent.changeText(emailInput, 'john.doe@gmail.com')

    await waitForExpect(() => {
      const enabledButtonSnapshot = toJSON()
      expect(disabledButtonSnapshot).toMatchDiffSnapshot(enabledButtonSnapshot)
    })
  })

  it('should redirect to ResetPasswordEmailSent when password reset request is successful', async () => {
    const { getByPlaceholderText, findByText } = renderPage()

    const emailInput = getByPlaceholderText('tonadresse@email.com')
    fireEvent.changeText(emailInput, 'john.doe@gmail.com')

    const validateEmailButton = await findByText('Valider')
    fireEvent.press(validateEmailButton)

    await waitForExpect(() => {
      expect(navigationTestProps.navigation.navigate).toBeCalledTimes(1)
      expect(navigationTestProps.navigation.navigate).toHaveBeenCalledWith(
        'ResetPasswordEmailSent',
        {
          email: 'john.doe@gmail.com',
        }
      )
    })
  })

  it('should redirect to Login when clicking on ArrowPrevious icon', async () => {
    const { getByTestId } = renderPage()

    const leftIcon = getByTestId('leftIcon')
    fireEvent.press(leftIcon)

    await waitForExpect(() => {
      expect(navigationTestProps.navigation.navigate).toBeCalledTimes(1)
      expect(navigationTestProps.navigation.navigate).toHaveBeenCalledWith('Login')
    })
  })
})

function renderPage() {
  return render(
    <ForgottenPassword
      {...(navigationTestProps as StackScreenProps<HomeStackParamList, 'ForgottenPassword'>)}
    />
  )
}
