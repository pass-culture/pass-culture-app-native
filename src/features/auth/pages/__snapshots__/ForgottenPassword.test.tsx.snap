// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ForgottenPassword /> should enable validate button when email input is filled 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -599,11 +599,11 @@
                    },
                  ]
                }
                textContentType=\\"emailAddress\\"
                underlineColorAndroid=\\"transparent\\"
-               value=\\"\\"
+               value=\\"john.doe@gmail.com\\"
              />
            </View>
          </View>
          <View
            numberOfSpaces={6}
@@ -615,22 +615,22 @@
              ]
            }
          />
          <View
            accessible={true}
-           focusable={false}
+           focusable={true}
            onClick={[Function onClick]}
            onResponderGrant={[Function onResponderGrant]}
            onResponderMove={[Function onResponderMove]}
            onResponderRelease={[Function onResponderRelease]}
            onResponderTerminate={[Function onResponderTerminate]}
            onResponderTerminationRequest={[Function onResponderTerminationRequest]}
            onStartShouldSetResponder={[Function onStartShouldSetResponder]}
            style={
              Object {
                \\"alignItems\\": \\"center\\",
-               \\"backgroundColor\\": \\"#ff99be\\",
+               \\"backgroundColor\\": \\"#eb0055\\",
                \\"borderRadius\\": 24,
                \\"borderWidth\\": 0,
                \\"flexDirection\\": \\"row\\",
                \\"height\\": 40,
                \\"justifyContent\\": \\"center\\","
`;
